### ISSUE TITLE
Pulumi forever stuck on preview

### ISSUE THREAD

**Issue by thecodeassassin** [pulumi/pulumi#6609]
## Expected behavior
A preview plan

## Current behavior
Gets stuck while running pulumi preview on google cloud build. Locally it works fine.

Using docker image: gcr.io/cloud-builders/go

output of TF_LOG=TRACE pulumi preview --logtostderr -v=9:
```
reviewing update (prod):
I0324 08:38:24.866006      56 plugins.go:77] gatherPluginsFromProgram(): gathering plugins from language host
I0324 08:38:24.866335      56 plugins.go:609] GetPluginPath(language, go, <nil>): found next to current executable /builder/home/.pulumi/bin/pulumi-language-go
I0324 08:38:24.866373      56 plugin.go:124] Launching plugin 'go' from '/builder/home/.pulumi/bin/pulumi-language-go' with args: 127.0.0.1:41159
I0324 08:38:24.876446      56 langruntime_plugin.go:178] langhost[go].GetPluginInfo() executing
I0324 08:38:24.877002      56 langruntime_plugin.go:91] langhost[go].GetRequiredPlugins(proj=mcls-streaming-infra,pwd=/workspace,program=.) executing
I0324 08:38:24.952025      56 langruntime_plugin.go:133] langhost[go].GetRequiredPlugins(proj=mcls-streaming-infra,pwd=/workspace,program=.) success: #versions=1
I0324 08:38:24.952061      56 plugins.go:89] gatherPluginsFromProgram(): plugin go <nil> () is required by language host
I0324 08:38:24.952087      56 plugins.go:89] gatherPluginsFromProgram(): plugin gcp 4.9.1 () is required by language host
I0324 08:38:24.952097      56 plugins.go:101] gatherPluginsFromSnapshot(): gathering plugins from snapshot
I0324 08:38:24.952105      56 plugins.go:104] gatherPluginsFromSnapshot(): no snapshot available, skipping
I0324 08:38:24.952112      56 plugins.go:133] ensurePluginsAreInstalled(): beginning
I0324 08:38:24.952333      56 plugins.go:627] GetPluginPath(resource, gcp, 4.9.1): enabling new plugin behavior
I0324 08:38:24.952354      56 plugins.go:709] SelectCompatiblePlugin(..., gcp): beginning
I0324 08:38:24.952362      56 plugins.go:748] SelectCompatiblePlugin(..., gcp): failed to find match
I0324 08:38:24.952478      56 plugins.go:609] GetPluginPath(language, go, <nil>): found next to current executable /builder/home/.pulumi/bin/pulumi-language-go
I0324 08:38:24.952837      56 plugins.go:138] ensurePluginsAreInstalled(): plugin go <nil> already installed
I0324 08:38:24.952695      56 plugins.go:146] ensurePluginsAreInstalled(): plugin gcp 4.9.1 not installed, doing install
I0324 08:38:24.953127      56 plugins.go:165] installPlugin(gcp, 4.9.1): beginning install
I0324 08:38:24.953232      56 plugins.go:172] installPlugin(gcp, 4.9.1): initiating download
I0324 08:38:24.953332      56 plugins.go:236] gcp downloading from https://get.pulumi.com/releases/plugins
I0324 08:38:24.953370      56 plugins.go:243] full plugin download url: https://get.pulumi.com/releases/plugins/pulumi-resource-gcp-v4.9.1-linux-amd64.tar.gz
I0324 08:38:24.953392      56 plugins.go:253] plugin install request headers: map[User-Agent:[pulumi-cli/1 (; linux)]]
I0324 08:38:25.071865      56 plugins.go:260] plugin install response headers: map[Accept-Ranges:[bytes] Age:[347394] Content-Disposition:[attachment; filename=pulumi-resource-gcp-v4.9.1-linux-amd64.tar.gz] Content-Length:[37015648] Content-Type:[application/x-gzip] Date:[Sat, 20 Mar 2021 08:08:32 GMT] Etag:["5062f2d4c63e32efb5765fe85f6b7a64-8"] Last-Modified:[Fri, 29 Jan 2021 18:20:31 GMT] Server:[AmazonS3] Via:[1.1 62c9e714a19291e1725b1320d4c9b369.cloudfront.net (CloudFront)] X-Amz-Cf-Id:[0NsUPh8s1VkDrRtfiIFTYB9YpZEhMTznVaDiDrVcJBZnCB_6POQGUw==] X-Amz-Cf-Pop:[ORD52-C2] X-Cache:[Hit from cloudfront]]
[resource plugin gcp-4.9.1] installing
I0324 08:38:25.072206      56 plugins.go:182] installPlugin(gcp, 4.9.1): extracting tarball to installation directory

Downloading plugin: 0 B / 35.30 MiB    0.00%
Downloading plugin: 3.97 MiB / 35.30 MiB   11.24% 1s
Downloading plugin: 8.29 MiB / 35.30 MiB   23.47% 1s
Downloading plugin: 12.05 MiB / 35.30 MiB   34.14% 1s
Downloading plugin: 16.07 MiB / 35.30 MiB   45.53%
Downloading plugin: 22.83 MiB / 35.30 MiB   64.66%
Downloading plugin: 34.12 MiB / 35.30 MiB   96.65%
Downloading plugin: 35.30 MiB / 35.30 MiB  100.00% 1s
I0324 08:38:26.332993      56 plugins.go:189] installPlugin(gcp, 4.9.1): successfully installed
I0324 08:38:26.333020      56 plugins.go:153] ensurePluginsAreInstalled(): completed
I0324 08:38:26.333057      56 plugins.go:241] computeDefaultProviderPlugins(): considering go
I0324 08:38:26.333072      56 plugins.go:244] computeDefaultProviderPlugins(): skipping go, not a resource provider
I0324 08:38:26.333079      56 plugins.go:241] computeDefaultProviderPlugins(): considering gcp-4.9.1
I0324 08:38:26.333092      56 plugins.go:272] computeDefaultProviderPlugins(): plugin gcp-4.9.1 selected for package gcp (first seen)
I0324 08:38:26.333153      56 plugins.go:278] computeDefaultProviderPlugins(): summary of default plugins:
I0324 08:38:26.333165      56 plugins.go:280]   gcp             = 4.9.1

I0324 08:38:26.333619      56 deployment_executor.go:400] deploymentExecutor.retirePendingDeletes(...): no pending deletions
I0324 08:38:26.333664      56 deployment_executor.go:221] deploymentExecutor.Execute(...): waiting for incoming events
I0324 08:38:26.333707      56 langruntime_plugin.go:144] langhost[go].Run(pwd=/workspace,program=.,#args=0,proj=mcls-streaming-infra,stack=prod,#config=4,dryrun=true) executing
I0324 08:38:26.333969      56 step_executor.go:327] StepExecutor worker(-2): worker coming online
I0324 08:38:26.334981      56 step_executor.go:327] StepExecutor worker(-2): worker waiting for incoming chains
```

**Comment by thecodeassassin** [comment]
I think the build steps fails:

```
     pulumi:pulumi:Stack  mcls-streaming-infra-prod           go build github.com/pulumi/pulumi-gcp/sdk/v4/go/gcp/compute: /usr/local/go/pkg/tool/linux_amd
I0324 10:31:36.084873    1823 langruntime_plugin.go:171] langhost[go].RunPlan(pwd=/srv,program=.,...,dryrun=true) success: progerr=program exited with non-zero exit code: 1
I0324 10:31:36.092400    1823 source_eval.go:178] EvalSourceIterator ended with an error: an unhandled error occurred: program exited with non-zero exit code: 1
I0324 10:31:36.095735    1823 deployment_executor.go:225] deploymentExecutor.Execute(...): incoming event (nil? true, &{0xc000616510})
I0324 10:31:36.096596    1823 eventsink.go:86] eventSink::Error(<{%reset%}>an unhandled error occurred: program exited with non-zero exit code: 1<{%reset%}>)
I0324 10:31:36.096769    1823 step_executor.go:327] StepExecutor worker(-1): StepExecutor.waitForCompletion(): waiting for worker threads to exit
I0324 10:31:36.097622    1823 step_executor.go:327] StepExecutor worker(-2): worker exiting due to cancellation
I0324 10:31:36.097709    1823 step_executor.go:327] StepExecutor worker(-1): StepExecutor.waitForCompletion(): worker threads all exited
I0324 10:31:36.097743    1823 deployment_executor.go:261] deploymentExecutor.Execute(...): step executor has completed
I0324 10:31:36.098760    1823 deployment_executor.go:132] deploymentExecutor.Execute(...): exiting provider canceller
I0324 10:31:36.101065    1823 ignore.go:44] Explicitly ignoring and discarding error: rpc error: code = Canceled desc = grpc: the client connection is closi
     pulumi:pulumi:Stack  mcls-streaming-infra-prod           1 error; 1 message

Diagnostics:
  pulumi:pulumi:Stack (mcls-streaming-infra-prod):
    error: an unhandled error occurred: program exited with non-zero exit code: 1

    go build github.com/pulumi/pulumi-gcp/sdk/v4/go/gcp/compute: /usr/local/go/pkg/tool/linux_amd64/compile: signal: killed
```

**Comment by komalali** [comment]
Can you provide us some code and steps to reproduce this issue?

**Comment by thecodeassassin** [comment]
Sure.

https://gist.github.com/thecodeassassin/5c0112bb1003c8ce74fe5f7e85936104 => main.go

https://gist.github.com/thecodeassassin/0b9b6c831a3ab2e89be1c2ce8aa7ccfa => networking.go

config:

https://gist.github.com/thecodeassassin/14c12e38d4a99102a0093d19d6e52a94

**Comment by thecodeassassin** [comment]
I also saw this happening today:

```
 go build github.com/pulumi/pulumi-gcp/sdk/v4/go/gcp/compute: /root/.go/pkg/tool/linux_amd64/compile: signal: killed
 ```
 
Could it be related?

**Comment by thecodeassassin** [comment]
We will be moving back to terraform because of this, it's a shame though.

**Comment by stack72** [comment]
Hey @thecodeassassin 

I'm sorry that this has been frustrating for you - TL;DR the go SDK in its current format has caused pain due to the size of the SDK it produces

We have been working in Pulumi-gcp v5.0.0-beta.1 (and Pulumi v3.0.0-beta.1) to make that SDK way better. We've actually reduced its size massively which means faster compilation and smaller binaries

Would you be interested in trying that for your use case?

Paul

**Comment by thecodeassassin** [comment]
Hi Paul, we already moved this project back to Terraform. Maybe we'll try again in the future. Thanks for responding.

**Comment by stack72** [comment]
No worries at all @thecodeassassin - sorry again this was a frustrating experience for you. I can assure you we do care about continually making things better

Paul

**Comment by paymog** [comment]
I'm running into this issue when running preview within a docker container. I'm running Pulumi v3.4 and pulumi gcp v5.2.0. Any ideas what might be going on @stack72? Are there workaround you can suggest?

**Comment by lukehoban** [comment]
@paymog the issue originally tracked by this issue was fixed in Pulumi 3.0, where build times and memory requirements were reduced by ~70%. What is the exact issue you are having now?  Might be good to open a new issue with any details you can share?

**Comment by paymog** [comment]
Got it, I just opened https://github.com/pulumi/pulumi/issues/7237
